File: src/2025-02-27-deepclaude37-rust/javascript/exercises/practice/robot-name/robot-name.js
// This is only a SKELETON file for the 'Robot Name' exercise. It's been
// provided as a convenience to get your started writing code faster.

export class Robot {
  static usedNames = new Set();
  #currentName = null;

  get name() {
    if (!this.#currentName) {
      let newName;
      do {
        newName = Robot.generateRandomName();
      } while (Robot.usedNames.has(newName));
      Robot.usedNames.add(newName);
      this.#currentName = newName;
    }
    return this.#currentName;
  }

  reset() {
    // Don't remove the old name from usedNames
    // This ensures all names ever generated remain unique
    this.#currentName = null;
  }

  static generateRandomName() {
    const letters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';
    const digits = '0123456789';
    
    let name = '';
    // Add two random uppercase letters
    for (let i = 0; i < 2; i++) {
      name += letters.charAt(Math.floor(Math.random() * letters.length));
    }
    
    // Add three random digits
    for (let i = 0; i < 3; i++) {
      name += digits.charAt(Math.floor(Math.random() * digits.length));
    }
    
    return name;
  }
}

Robot.releaseNames = () => {
  Robot.usedNames.clear();
};

File: src/python-2025-02-23-yh-deepclaude-retry/python/exercises/practice/robot-name/robot_name.py
import random

class Robot:
    _used_names = set()

    def __init__(self):
        self._name = None
        self._generate_name()

    @property
    def name(self):
        return self._name

    def _generate_name(self):
        while True:
            letters = ''.join(random.choices('ABCDEFGHIJKLMNOPQRSTUVWXYZ', k=2))
            digits = ''.join(random.choices('0123456789', k=3))
            new_name = letters + digits
            if new_name not in Robot._used_names:
                Robot._used_names.add(new_name)
                self._name = new_name
                break

    def reset(self):
        if self._name in Robot._used_names:
            Robot._used_names.remove(self._name)
        self._generate_name()
